sqlsimplekv/keyvalue.go

```diff
@@ -0,0 +1,214 @@
+type kvStore struct {
+	db     *sql.DB
+	driver *driver
+	store  string
+}
+
+// Context implements simplekv.Store.Context.
+func (s *kvStore) Context(ctx context.Context) (context.Context, func()) {
+	return ctx, func() {}
+}
+
+type keyValueParams struct {
+	argBuilder
+
+	Store  string
+	Key    string
+	Value  []byte
+	Expire nullTime
+	Update bool
+}
+
+// Get implements simplekv.Store.Get by selecting the blob with the
+// given key from the provider_data table.
+func (s *kvStore) Get(ctx context.Context, key string) ([]byte, error) {
+	v, err := s.get(s.db, key, false)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(simplekv.ErrNotFound))
+	}
+	return v, nil
+}
+
+// get is like Get except that it operates on a general queryer value.
+func (s *kvStore) get(q queryer, key string, forUpdate bool) ([]byte, error) {
+	params := &keyValueParams{
+		argBuilder: s.driver.argBuilderFunc(),
+		Store:      s.store,
+		Key:        key,
+	}
+	var value []byte
+	tmpl := tmplGetKeyValue
+	if forUpdate {
+		tmpl = tmplGetKeyValueForUpdate
+	}
+	row, err := s.driver.queryRow(q, tmpl, params)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	if err := row.Scan(&value); err != nil {
+		if errgo.Cause(err) == sql.ErrNoRows {
+			return nil, simplekv.KeyNotFoundError(key)
+		}
+		return nil, errgo.Mask(err)
+	}
+	return value, nil
+}
+
+// Set implements simplekv.Store.Set by upserting the blob with the
+// given key, value and expire time into the provider_data table.
+func (s *kvStore) Set(ctx context.Context, key string, value []byte, expire time.Time) error {
+	return s.set(s.db, key, value, expire)
+}
+
+// set is like Set except that it operates on a general queryer value.
+func (s *kvStore) set(q queryer, key string, value []byte, expire time.Time) error {
+	params := &keyValueParams{
+		argBuilder: s.driver.argBuilderFunc(),
+		Store:      s.store,
+		Key:        key,
+		Value:      value,
+		Expire:     nullTime{expire, !expire.IsZero()},
+		Update:     true,
+	}
+	_, err := s.driver.exec(q, tmplInsertKeyValue, params)
+	return errgo.Mask(err)
+}
+
+// Update implements simplekv.Store.Update.
+func (s *kvStore) Update(ctx context.Context, key string, expire time.Time, getVal func(old []byte) ([]byte, error)) error {
+	err := s.withTx(func(tx *sql.Tx) error {
+		v, err := s.get(tx, key, true)
+		if err != nil {
+			if errgo.Cause(err) != simplekv.ErrNotFound {
+				return errgo.Mask(err)
+			}
+		} else if v == nil {
+			v = []byte{}
+		}
+		newVal, err := getVal(v)
+		if err != nil {
+			return errgo.Mask(err, errgo.Any)
+		}
+		if err := s.set(tx, key, newVal, expire); err != nil {
+			return errgo.Mask(err)
+		}
+		return nil
+	})
+	return errgo.Mask(err, errgo.Any)
+}
+
+// Add implements simplekv.Store.Add by inserting a blob with the
+// given key, value and expire time into the provider_data table.
+func (s *kvStore) Add(_ context.Context, key string, value []byte, expire time.Time) error {
```

Comment:
This function isn't in the interface any more.

Reply:
done